FastSimulation/Tracking/src/SeedFinderSelector.cc

```diff
@@ -119,8 +154,140 @@ bool SeedFinderSelector::pass(const std::vector<const FastTrackerRecHit *>& hits
+      	  return !tripletresult.empty();
+	}
     }
+    //new for Phase1     
+    if(CAHitQuadGenerator_)
+    {
+      if(hits.size() < 4)
+	{
+	  throw cms::Exception("FastSimTracking") << "For the given configuration, SeedFinderSelector::pass requires at least 4 hits";
+	}
+
+      if(!seedingLayer)
+	throw cms::Exception("FastSimTracking") << "ERROR: SeedingLayers pointer not set for CAHitQuadrupletGenerator";      
+
+      SeedingLayerSetsHits & layers = *seedingLayer;
+      //constructing IntermediateHitDoublets to be passed onto CAHitQuadrupletGenerator::hitNtuplets()
+      IntermediateHitDoublets ihd(&layers);
+      const TrackingRegion& tr_ = *trackingRegion_;
+      auto filler = ihd.beginRegion(&tr_);
+      
+      //Forming doublets for CA quadruplets from the allowed layer pair combinations:(0,1),(1,2),(2,3)
+      std::string hitPair[2];
+      for(int i=0; i<3; i++){
+	SeedingLayerSetsHits::SeedingLayerSet pairCandidate;
+	hitPair[0] = Layer_name(hits[i]);
+	hitPair[1] = Layer_name(hits[i+1]);
+       
+	bool found;
+        for(SeedingLayerSetsHits::SeedingLayerSet ls : *seedingLayer){
+	  found = false;
+	  for(const auto p : layerPairs_){
+	    pairCandidate = ls.slice(p,p+2);
+	    if(hitPair[0] == pairCandidate[0].name() && hitPair[1] == pairCandidate[1].name()){
+	      found = true;
+	      break;
+	    }
+	  }
+	  if(found)
+	    break;
+	}
+	assert(found == true);
+	const DetLayer * fLayer = measurementTracker_->geometricSearchTracker()->detLayer(hits[i]->det()->geographicalId());
+	const DetLayer * sLayer = measurementTracker_->geometricSearchTracker()->detLayer(hits[i+1]->det()->geographicalId());
+	std::vector<BaseTrackerRecHit const *> fHits{hits[i]};
+	std::vector<BaseTrackerRecHit const *> sHits{hits[i+1]};
+
+	//Important: doublets to be added to the cache
+	auto& layerCache = filler.layerHitMapCache();
+	const RecHitsSortedInPhi& firsthm = *layerCache.add(pairCandidate[0], std::make_unique<RecHitsSortedInPhi>(fHits, trackingRegion_->origin(), fLayer));
+	const RecHitsSortedInPhi& secondhm = *layerCache.add(pairCandidate[1], std::make_unique<RecHitsSortedInPhi>(sHits, trackingRegion_->origin(), sLayer));
+	HitDoublets res(firsthm,secondhm);
+	HitPairGeneratorFromLayerPair::doublets(*trackingRegion_,*fLayer,*sLayer,firsthm,secondhm,*eventSetup_,0,res);
+	filler.addDoublets(pairCandidate, std::move(res)); //fill the formed doublet
+      }
+      
+      std::vector<OrderedHitSeeds> quadrupletresult;
+      quadrupletresult.resize(ihd.regionSize());
+      for(auto& ntuplet : quadrupletresult)
+	ntuplet.reserve(4);
+      CAHitQuadGenerator_->hitNtuplets(ihd,quadrupletresult,*eventSetup_,*seedingLayer); //calling the function from the class, modifies quadrupletresult
+      return !quadrupletresult.empty();  
+    }    
+
     return true;
     
 }
+
+//new for Phase1
+std::string SeedFinderSelector::Layer_name(const FastTrackerRecHit * hit) const
```

Comment:
this part differs from what's in #23363 
Please provide a brief explanation (either in respoonse to this message or in the PR description).
Do we have an equivalent for non-fastsim code to check that this pattern is used in production code already?

Reply:
as I commented above, the change is necessary because of other tracking code updates that aren't in 94X. Maybe @angirar or @makortel can provide more details.

Reply:
On a cursory look the changes needed in the last commit (7fc2cc259f657211e1e7dcb20eb16ddd84cd4e68) could be caused by #21138 being merged in 10_0_0_pre1 (so those developments not being available in 9_4).

Reply:
Thank you @makortel , that is precisely the reason why I had to implement these new changes. The strategy is still the same as followed in the original PR #23363, only difference is that I check for the "std::string" layer name of the hit instead of the "SeedingLayerId = std::tuple<GeomDetEnumerators::SubDetector, TrackerDetSide, int>" of the layer.